runtime/src/main/java/org/corfudb/runtime/view/StreamsView.java

```diff
@@ -131,91 +131,105 @@ public void gc(long trimMark) {
-
-        ld.checkMaxWriteSize(runtime.getParameters().getMaxWriteSize());
-
         TokenResponse tokenResponse = null;
-        for (int x = 0; x < runtime.getParameters().getWriteRetry(); x++) {
-            // Go to the sequencer, grab a token to write.
-            tokenResponse = conflictInfo == null
-                    ? runtime.getSequencerView().next(streamIDs) // Token w/o conflict info
-                    : runtime.getSequencerView().next(conflictInfo, streamIDs); // Token w/ conflict info
-
-            // Is our token a valid type?
-            AbortCause abortCause = null;
-            switch (tokenResponse.getRespType()) {
-                case TX_ABORT_CONFLICT:
-                    abortCause = AbortCause.CONFLICT;
-                    break;
-                case TX_ABORT_NEWSEQ:
-                    abortCause = AbortCause.NEW_SEQUENCER;
-                    break;
-                case TX_ABORT_SEQ_OVERFLOW:
-                    abortCause = AbortCause.SEQUENCER_OVERFLOW;
-                    break;
-                case TX_ABORT_SEQ_TRIM:
-                    abortCause = AbortCause.SEQUENCER_TRIM;
-                    break;
-            }
 
-            if (abortCause != null) {
-                throw new TransactionAbortedException(
-                        conflictInfo,
-                        tokenResponse.getConflictKey(), tokenResponse.getConflictStream(),
-                        tokenResponse.getToken().getSequence(), abortCause,
-                        TransactionalContext.getCurrentContext());
-            }
-
-            try {
-                if (TransactionalContext.isInTransaction()) {
-                    // If this transaction has entries that wish to capture the committed address
-                    // invoke its preCommitCallbacks with the tokenResponse from the sequencer.
-                    // Note that we might invoke the same method multiple times on retries,
-                    // which means the preCommitCallback must be idempotent.
-                    TokenResponse finalTokenResponse = tokenResponse;
-                    log.debug("append: Invoking {} preCommitListeners",
-                            TransactionalContext.getRootContext().getPreCommitListeners().size());
-                    TransactionalContext.getRootContext()
-                            .getPreCommitListeners()
-                            .forEach(e -> e.preCommitCallback(finalTokenResponse));
+        // Opening serialization handle before acquiring token, this way we prevent the
+        // readers to wait for the possibly long serialization time in writer.
+        // The serialization here only serializes the payload because the token is not
+        // acquired yet, thus metadata is incomplete. Once a token is acquired, the
+        // writer will append the serialized metadata to the buffer.
+        try (ILogData.SerializationHandle sh = ld.getSerializedForm(serializeMetadata)) {
```

Comment:
the whole method is quite big, can you split it to a couple of smaller ones?

Reply:
Done.